src/ooo_cpu.cc

```diff
@@ -304,11 +304,58 @@ void O3_CPU::fetch_instruction()
       return;
     }
 
-  // scan through IFETCH_BUFFER to find instructions that need to be translated
+  std::size_t available_fetch_bandwidth = FETCH_WIDTH;
+
+  // scan through IFETCH_BUFFER to find instructions that hit in the decoded instruction buffer
   uint32_t index = IFETCH_BUFFER.head;
   for(uint32_t i=0; i<IFETCH_BUFFER.SIZE; i++)
     {
-        ooo_model_instr &ifb_entry = IFETCH_BUFFER.entry[index];
+      ooo_model_instr &ifb_entry = IFETCH_BUFFER.entry[index];
+
+      // Check DIB to see if we recently fetched this line
+      dib_t::value_type &dib_set = DIB[(ifb_entry.ip >> LOG2_BLOCK_SIZE) % DIB_SET];
+      auto way = std::find_if(dib_set.begin(), dib_set.end(), [ifb_entry](dib_entry_t x){ return x.valid && ((x.addr >> LOG2_BLOCK_SIZE) == (ifb_entry.ip >> LOG2_BLOCK_SIZE));});
```

Comment:
I don't think I agree that we ought to be searching for the block address instead of the byte address. Since the DIB "stores micro-ops" (it doesn't, but...), it feels to me like an entry should correspond to a single instruction, not a whole line that may contain many instructions.

There might be a case to be made that if the block address hits, we can skip translation, or possibly translation+fetch, but I don't think a block address hit should cause us to skip decode.

Reply:
The idea of the DIB is that it provides a similar benefit to a modern uop cache, which in the case of the latest Intel core is reported to cache ~2.25k instructions.  A real uop cache stores decoded instructions on the granularity of chunks of several instructions, and not individual instructions or whole cache lines.

If each entry of the DIB stores only a single instruction, then we're only covering 64 instructions.  This is off of our ~2.25k instruction target by a factor of 32.  As a compromise, in our previous conversation on the topic we had decided that the DIB would cache entire cache lines, which can be assumed to hold on average 32 instructions each, getting us all the way to our ~2.25k target.

Reply:
I see. That makes a lot of sense. Since we aren't attempting any kind of uop decoding, cache lines are a fair analog for instruction windows.